compiler/src/iree/compiler/Codegen/Common/ConvertUnsupportedFloatArithPass.cpp

```diff
@@ -0,0 +1,107 @@
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===--------------- ConvertUnsupportedFloatArithPass.cpp ----------------===//
+//
+//   Emulate arith and vector floating point operations that use float types
+//   which are unspported on a target by inserting extf/truncf pairs around all
+//   such operations in order to produce arithmetic that can be performed while
+//   preserving the original rounding behavior.
+//
+//===---------------------------------------------------------------------===//
+
+#include "iree/compiler/Codegen/Common/Passes.h"
+#include "iree/compiler/Codegen/Common/Transforms.h"
+#include "iree/compiler/Codegen/Utils/GPUUtils.h"
+#include "mlir/Dialect/AMDGPU/Utils/Chipset.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Arith/Transforms/Passes.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/Interfaces/FunctionInterfaces.h"
+
+namespace mlir::iree_compiler {
+
+#define GEN_PASS_DEF_CONVERTUNSUPPORTEDFLOATARITHPASS
+#include "iree/compiler/Codegen/Common/Passes.h.inc"
+
+namespace {
+
+struct ConvertUnsupportedFloatArithPass final
+    : public impl::ConvertUnsupportedFloatArithPassBase<
+          ConvertUnsupportedFloatArithPass> {
+  void runOnOperation() override;
+  using Base::Base;
+};
+
+} // namespace
+//
+
+// Populates source and target conversion types based on the target
+// architecture.
+// TODO(pashu123): Refine the patterns based on the target arch.
+static LogicalResult
+populateSourceAndTargetType(MLIRContext *ctx, Operation *op,
+                            SmallVectorImpl<Type> &sourceTypes,
+                            Type &targetType) {
+
+  auto gpuAttr = getGPUTargetAttr(op);
+  if (!gpuAttr)
+    return failure();
+
+  StringRef chipset = gpuAttr.getArch();
+  FailureOr<amdgpu::Chipset> maybeChipset = amdgpu::Chipset::parse(chipset);
+  if (failed(maybeChipset)) {
+    op->emitOpError() << "Invalid chipset name: " << chipset;
+    return failure();
+  }
+  // Add source and target conversion types for gfx94{*} series.
+  if (maybeChipset->majorVersion == 9 && maybeChipset->minorVersion == 4) {
+    sourceTypes.insert(sourceTypes.end(), {Float8E4M3FNUZType::get(ctx),
+                                           Float8E5M2FNUZType::get(ctx)});
+    targetType = Float32Type::get(ctx);
+    return success();
+  }
+  return failure();
+}
+
+void ConvertUnsupportedFloatArithPass::runOnOperation() {
+
+  MLIRContext *context = &getContext();
+  Operation *op = getOperation();
+  SmallVector<Type> sourceTypes;
+  Type targetType;
+
+  if (failed(
+          populateSourceAndTargetType(context, op, sourceTypes, targetType))) {
+    return;
+  }
+
+  if (sourceTypes.empty()) {
+    (void)emitOptionalWarning(
+        std::nullopt,
+        "no source types specified, float emulation will do nothing");
+    return signalPassFailure();
+  }
```

Comment:
Doesn't it mean that it is a nop pass? In this case, we should not signal a pass failure.

Reply:
Any best practice in this case? Or a simple `return;`

Reply:
I'd suggest a log to LLVM_DEBUG and just a simple `return;`.

Reply:
I see. Thanks.